database.py

```diff
@@ -1,14 +1,20 @@
 import sqlite3
+from typing import TYPE_CHECKING, Iterator, Optional
 
 import config
 
+if TYPE_CHECKING:
+    from lib import RunResult
```

Comment:
How do we not have a circular import here?

Reply:
typing.TYPE_CHECKING is a special constant that is always False at runtime, but True to typecheckers, this lets us type circular things without actually causing a circular import

Reply:
So when type-checking a file, you don't have to be able to have the entire app of files imported at the same time?

Reply:
type checkers use similar magic to what rustc does when compiling crates, i/e they group every file into one big AST soup and then resolve stuff as needed, instead of doing what python does where it loads the file you pass it, and then imports stuff as you write `import foo` linearly